pkg/nsx/services/firewall.go

```diff
@@ -1322,58 +1361,135 @@ func (service *SecurityPolicyService) CreateOrUpdateSecurityPolicy(obj *v1alpha1
 	for _, rule := range indexResults {
 		existingRules = append(existingRules, rule.(model.Rule))
 	}
-	if service.securityPolicyEqual(&existingSecurityPolicy, nsxSecurityPolicy) && service.rulesEqual(existingRules, nsxSecurityPolicy.Rules) {
-		log.Info("security policy not changed, skipping", "nsxSecurityPolicy.Id", nsxSecurityPolicy.Id)
+	// Caution! createOrUpdate can't delete the legacy rules.
+	// So we have to delete the legacy rules firstly by rulesClient.
+	spEqual := service.securityPolicyEqual(&existingSecurityPolicy, nsxSecurityPolicy)
+	ruleEqual, legacyRules := service.rulesEqual(existingRules, nsxSecurityPolicy.Rules)
+	if spEqual && ruleEqual {
+		log.Info("security policy and rules not changed, skip updating them", "nsxSecurityPolicy.Id", nsxSecurityPolicy.Id)
 	} else {
-		err = policiesClient.Patch(service.getDomain(), *nsxSecurityPolicy.Id, *nsxSecurityPolicy)
+		err := service.updateOrDeleteRules(nsxSecurityPolicy, legacyRules)
```

Comment:
1. The store update should happen after you have successfully patched the NSX resource.
2. You update the NSX rules in `createOrUpdateSecurityPolicy` but delete the old rules in `updateOrDeleteRules`, which puts the actions of the NSX rules in 2 different places (code levels) and introduces logical confusion. My suggestion is to rename `updateOrDeleteRules` to `updateOrDeleteRulesInStore` and introduce a new function `DeleteRules` to delete old rules like you did with old groups.

Reply:
OK